### DISCUSSION TITLE
Can GPU be used for automatic fiber bundle extraction?

### DISCUSSION THREAD

**Discussion by Sunita2023** [dipy/dipy#3406] [Tractography / Tractometry]
Dear Jong,

Is their a way to use GPU in automatic fiber bundle extraction?

Thanks in advance.

Best,
Sunita

**Comment by arokem** [comment]
Hello! We have an implementation of GPU-accelerated tractography, which should make a key part of the process a lot faster: https://github.com/dipy/GPUStreamlines

**Reply by Sunita2023** [reply]
Thank you so much Ariel. 
The link explains about GPU-accelerated tractography. However, recognize and refinement also take a lot of time. Is it possible to accelerate the process.   

Best,
Sunita

**Comment by Garyfallidis** [comment]
Yes, it is definitely possible @Sunita2023 and there is plenty of upcoming work on this topic. We will need to wait for the papers to be published to speak openly about this. Maybe others here can provide more insights. But if you want to get going faster with this you may need to make a research collaboration request (by email). Stating what problem you are trying to solve etc. I hope this helps.

**Reply by Sunita2023** [reply]
Dear Prof Garyfallidis,

First of all I would like to thank for the amazing work you are contributing to the world through your research.  Your work has tremendous practical implication specially in clinical field. I have read your papers and watched your videos on Dipy. Recently, we have started learning and using Dipy to process dMRI. Still there are so many things yet to be learnt. 

I have few queries. 
For registration why Dipy takes so long time where other available softwares are quite fast?

Using FURY, can we show internal anatomy of 3D brain in a scrollable manner in any plane with fiber track and other ROIs in 3D?

There are a number of methods (model, model free, CSD) to generate streamlines. Research papers suggest that CSD model provides the best results. For each model, we can use either probabilistic or deterministic approach to generate streamline. Am I right? 

We would like to test some of these methods which work better. It is bit confusing which models to use for testing purpose. It takes lot of time for refinement.  

I have confusion here as well. For example in Q-ball model, CSA is used as a model (https://docs.dipy.org/stable/examples_built/reconstruction/reconst_csa.html) while in CSD, CSA is used as a stopping criteria (https://docs.dipy.org/stable/examples_built/fiber_tracking/tracking_probabilistic.html#sphx-glr-examples-built-fiber-tracking-tracking-probabilistic-py). In Multi-shell multi-tissue CSD, qball_model is used as a model. Which one is the correct implementation of qball model?

I would really appreciate the help and support we receive from this forum. 

Wishing you all the very best for all your future endeavors.  

Thank you so much.

Warm regards,
Sunita

**Comment by Garyfallidis** [comment]
Hi @Sunita2023,
Thank you for your kind works.
Let's first try to understand why registration takes time for you.
What registration algorithm are you using, what are the input data and parameters?
Also, what have you used so far that was faster than what is available in DIPY?

**Reply by Sunita2023** [reply]
Hi Prof Garyfallidis, Affine registration using Simple ITK is much faster than DIPY.  It is just a small problem. Major problem is refinement which is taking 7-8 hrs for filtering streamlines generated from DTI tensor model. Now I am running CSD model. Looks like it will take 8 to 10 hours. Unless it is faster it is difficult to use. We are using recobundle as we want best refinement. Testing few other models does not seem to be feasible.  I will communicate via email if it is possible to use GPU version for the refinement. Thank you so much.

**Comment by Garyfallidis** [comment]
Dear @Sunita2023 thank your for the feedback and criticism. Note that in DIPY, efficiency is one of our primary goals. We are currently working to make DIPY faster than anything else out there. 

Nonetheless, DIPY's image registration modules are as fast as Ants (on a single core). We do not have CPU parallelization as of yet for image registration. We do have GPU acceleration for nonrigid. 

If refinement is taking 7-8 hours this probably means that you are doing something wrong on your side.
Similarly CSD should be done in 30' max. 8-10 hours for CSD is unlikely.

In summary, my guess is that you are doing something wrong in the way you are running things. Or you have some installation issues.
**Please provide an example dataset and the scripts that you are currently using so we can give you an informed answer.**

We have processed tens of thousands of datasets with Recobundles ... so I am confident to say that the issue is somewhere else. You may have an issue with too many threads opening. I do remember that multiple people had such an issue that ended up increasing runtimes across multiple DIPY modules.

Thank you.

**Reply by Sunita2023** [reply]
Dear prof. Garyfallidis, I have sent an email to your gmail account with all the required files. Thank you so much.
